engines/experiment/docs/rpc_plugins.md

```diff
@@ -0,0 +1,398 @@
+implementation. 
+
+First, clone the Turing repo and change the workdir to the reference plugin root directory: 
+```shell
+$ git clone https://github.com/gojek/turing.git
+$ cd turing/engines/experiment/examples/plugins/hardcoded
+```
+
+A Go plugin is a standalone application which serves the Experiment Engine implementation by calling 
+`rpc.Serve(&rpc.ClientServices{})`. The entrypoint of the application is [`cmd/main.go`](../examples/plugins/hardcoded/cmd/main.go):
+
+```go
+package main
+
+import (
+	"github.com/gojek/turing/engines/experiment/examples/plugins/hardcoded"
+	"github.com/gojek/turing/engines/experiment/plugin/rpc"
+)
+
+func main() {
+	rpc.Serve(&rpc.ClientServices{
+		Manager: &hardcoded.ExperimentManager{},
+		Runner:  &hardcoded.ExperimentRunner{},
+	})
+}
+```
+Note that the main function is intentionally simplified to just a call to `rpc.Serve` function. It's entirely 
+possible, that for the more complex plugin implementations it would be required to do some extra setup of the plugin 
+before it gets served. 
+
+All the bootstrapping code, required to serve the Experiment Engine plugin via the RPC, is provided by the Turing
+library, which is added into the plugin module as the dependency in [`go.mod`](../examples/plugins/hardcoded/go.mod):
+```shell
+module github.com/gojek/turing/engines/experiment/examples/plugins/hardcoded
+
+go 1.14
+
+require github.com/gojek/turing/engines/experiment v1.0.0
+
+replace github.com/gojek/turing/engines/experiment => ../../../
+```
+NOTE: Since this example plugin module and the `engines/experiment` library belong to the same repository, the plugin
+resolves `engines/experiment` package locally and that's the reason why the `replace ...` line is added to the 
+bottom of `go.mod`.
+
+A plugin needs to serve the implementation of [Experiment Manager](./developer_guide.md#experiment-manager) (either 
+[Standard](./developer_guide.md#standard-experiment-manager) or [Custom](./developer_guide.md#custom-experiment-manager)) 
+and [Experiment Runner](./developer_guide.md#experiment-runner). 
+
+#### Experiment Manager
+
+Both Standard and Custom experiment managers should implement the `ConfigurableExperimentManager` interface of:
+```go
+// ConfigurableExperimentManager interface of an ExperimentManager, that can be configured
+// with an arbitrary configuration passed as a JSON data
+type ConfigurableExperimentManager interface {
+    shared.Configurable
+    manager.ExperimentManager
+}
+```
+
+Which is composed of the `ExperimentManager` interface of:
+```go
+type ExperimentManager interface {
+    // GetEngineInfo returns the configuration of the experiment engine
+    GetEngineInfo() (Engine, error)
+    
+    // ValidateExperimentConfig validates the given Turing experiment config for the expected data and format
+    ValidateExperimentConfig(cfg json.RawMessage) error
+    
+    // GetExperimentRunnerConfig converts the given config (as retrieved from the DB) into a format suitable
```

Comment:
Another question here just to understand this more deeply, if `GetExperimentRunnerConfig` is only called at the time of router deployment, would the router require a re-deployment if the underlying experiment variable for the Router changes (e.g addition of a new segmenter)? If not, how would the update be propagated?

Reply:
To understand your question better, let me ask if this is what you mean:

1. There are some experiments configured in the experiment engine
2. Turing router, that uses this experiment engine is deployed
3. Experiments in the experiment engine are updated (i.e experiment variables, segmenters etc.)

4. ???

If this is the scenario you are talking about, then yes – it's expected that the new version of the router must be created and the configuration on how to extract the new variable from the request should be added into this version.

Reply:
Yes that's what I meant - thanks for clarifying! Would this above workflow description be of value to add to the documentation?

Reply:
This documentation is expected to cover primarily the development process of plugins. The general information of the experiment engine and its contract was expected to be in [developer_guide.md](https://github.com/gojek/turing/blob/main/engines/experiment/docs/developer_guide.md). I don't mind adding it into either of the two docs, but it feels like an isolated chunk of the documentation, which isn't very useful without the overhauling of the [developer_guide.md](https://github.com/gojek/turing/blob/main/engines/experiment/docs/developer_guide.md)

Reply:
👌🏻  We can park that for future dev_guide documentation improvement then.

Reply:
cc @krithika369

Reply:
I somehow feel this may be better in the user workflow doc than dev (the user needs to remember to update the router deployment if critical experiment config that affects the run experiment call, no matter the engine, changes). If the engine used is the in-house Turing Experiments, we can probably help them by having some advanced cross-app checks and balances (may be in the long term roadmap) and alert the user. But I think such changes may be infrequent and kind of obvious to the user. So I'm not too concerned.